### PR TITLE
JP-1440: Make source catalog background estimation more robust

### LINKED ISSUE #4913: source_catalog failing for some MIRI and NIRCam images
_Issue [JP-1440](https://jira.stsci.edu/browse/JP-1440) was created on JIRA by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse):_

DMS reports an error being thrown by the calwebb_image3 source_catalog step for some MIRI and NIRCam datasets using the latest 0.16.0 (B7.5rc1) delivery in their test environment. Traceback from a MIRI example is:
```java
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() CRDS version: 7.4.1.3, b7.4.1.3, b38c37e3437101250cbf746e1acc2fbfb34a8005 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() server: https://jwst-serverless-mode.stsci.edu 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() observatory: jwst 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() reftypes: ('abvegaoffset',) 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() context: None 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() ignore_cache: False 
2020-05-08 05:12:39,792 - CRDS - DEBUG - CRDS_PATH = ████████████████ 
2020-05-08 05:12:39,792 - CRDS - DEBUG - CRDS_SERVER_URL = https://jwst-serverless-mode.stsci.edu 
2020-05-08 05:12:39,793 - CRDS - DEBUG - Final effective context is 'jwst_0614.pmap' 
2020-05-08 05:12:39,793 - CRDS - DEBUG - Computing best references locally. 
2020-05-08 05:12:39,794 - CRDS - DEBUG - Bestrefs header: 
2020-05-08 05:12:39,794 - CRDS - DEBUG - {'META.EXPOSURE.READPATT [READPATT]': 'FAST', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.EXPOSURE.TYPE [EXP_TYPE]': 'MIR_IMAGE', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.BAND [BAND]': 'UNDEFINED', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.CHANNEL [CHANNEL]': 'UNDEFINED', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.CORONAGRAPH [CORONMSK]': 'UNDEFINED', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.DETECTOR [DETECTOR]': 'MIRIMAGE', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.FILTER [FILTER]': 'F770W', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.LAMP_STATE [LAMP]': 'OFF', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.NAME [INSTRUME]': 'MIRI', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.OBSERVATION.DATE [DATE-OBS]': '2016-01-17', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.OBSERVATION.TIME [TIME-OBS]': '01:04:27.879000', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.SUBARRAY.NAME [SUBARRAY]': 'SUB128', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.VISIT.TSOVISIT [TSOVISIT]': 'F', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'REFTYPE': 'UNDEFINED'} 
2020-05-08 05:12:39,795 - CRDS - DEBUG - Reference type 'abvegaoffset' defined as 'jwst_miri_abvegaoffset_0001.asdf' 
2020-05-08 05:12:39,795 - stpipe.Image3Pipeline.source_catalog - INFO - Using ABVEGAOFFSET reference file ████████████████████████████████████████████████████████████████ 
2020-05-08 05:12:39,795 - stpipe.Image3Pipeline.source_catalog - INFO - Instrument: MIRI 
2020-05-08 05:12:39,795 - stpipe.Image3Pipeline.source_catalog - INFO - Detector: MIRIMAGE 
2020-05-08 05:12:39,795 - stpipe.Image3Pipeline.source_catalog - INFO - Filter: F770W 
2020-05-08 05:12:39,796 - stpipe.Image3Pipeline.source_catalog - INFO - Subarray: SUB128 
Traceback (most recent call last): 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/bin/strun", line 26, in <module> 
step = Step.from_cmdline(sys.argv[1:]) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 156, in from_cmdline 
return cmdline.step_from_cmdline(args) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/cmdline.py", line 346, in step_from_cmdline 
step.run(*positional) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run 
step_result = self.process(*args) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/pipeline/calwebb_image3.py", line 111, in process 
self.source_catalog(result) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run 
step_result = self.process(*args) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog_step.py", line 76, in process 
model.data -= bkg.background 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/astropy/utils/decorators.py", line 758, in __get__ 
val = self.fget(obj) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog.py", line 334, in background 
return self._background2d.background 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/astropy/utils/decorators.py", line 758, in __get__ 
val = self.fget(obj) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog.py", line 321, in _background2d 
sigma_clip=sigma_clip, bkg_estimator=bkg_estimator) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/background/background_2d.py", line 338, in __init__ 
self._calc_bkg_bkgrms() 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/background/background_2d.py", line 671, in _calc_bkg_bkgrms 
idx = self._select_meshes(data_sigclip) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/background/background_2d.py", line 455, in _select_meshes 
self.exclude_percentile)) 
ValueError: All meshes contain > 1000.0 (10.0 percent per mesh) masked pixels. Please check your data or decrease "exclude_percentile". 
---------------------------------------------------------------------- 
ERROR RUNNING STEP 'Image3Pipeline': 
All meshes contain > 1000.0 (10.0 percent per mesh) masked pixels. 
Please check your data or decrease "exclude_percentile". 
----------------------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by larrybradley**
This PR adds a fallback background estimation if the background cannot be determined in meshes.  This PR also adds some checks for completely-masked data.

Fixes #4913 / [JP-1440](https://jira.stsci.edu/browse/JP-1440)

### CODE REVIEW

**Comment by larrybradley** [comment]
@hbushouse I added the changelog entry.  Note this PR also fixes the NIRISS issue you reported.

File: CHANGES.rst [hunk 1]
```diff
@@ -163,6 +163,9 @@ source_catalog
 
 - Updated step arguments in the documentation. [#4723]
 
+- Added fallback background estimation method to make background
```
Comment (hbushouse) [inline CHANGES.rst:166]: This should go into a new source_catalog entry up at the top of the file, under version 0.16.1 (not 0.16.0, which was already released).

Comment (larrybradley) [reply CHANGES.rst]: Fixed, thanks!

**Comment by larrybradley** [comment]
This PR also fixes the nearest-neighbor code when exactly one source is detected (also reported in JP-1440)

**Pull request merged**